src/ast.rs

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 VMware, Inc.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+use serde_json::Value;
+
+/// A path is a tree of selector nodes.
+///
+/// For example, the JSONPath `$.foo.bar` yields this AST:
+///
+/// ```text
+///               ^
+///              / \
+///             ^   \___ DotName("bar")
+///            / \
+///           ^   \___ DotName("foo")
+///          /
+/// Root ___/
+/// ```
+///
+/// A more complicated example: `$.foo[1,2]["bar"]`:
+///
+/// ```text
+///                  ^
+///                 / \
+///                ^   \___ Union
+///               / \            \
+///              ^   \___ Union   \
+///             /            \     [Field("bar")]
+///            ^              \
+///           / \              [Number(1), Number(2)]
+///          /   \
+/// Root ___/     \___ DotName("foo")
+/// ```
+///
+#[derive(Debug)]
+pub enum Path {
+    Root,
+    Sel(Box<Path>, Selector),
+}
+
+#[derive(Debug)]
+pub enum Selector {
+    Union(Vec<Index>),
+    DotName(String),
+    DotWildcard,
+}
+
+#[derive(Debug)]
+pub enum Index {
+    Field(String),
+    Number(i64),
+}
+
+type Iter<'a> = Box<dyn Iterator<Item = &'a Value> + 'a>;
+
+impl Path {
+    pub fn find<'a>(&'a self, input: &'a Value) -> Iter<'a> {
+        match self {
+            Path::Root => Box::new(std::iter::once(input)),
+            Path::Sel(left, sel) => Box::new(left.find(input).flat_map(move |v| sel.find(v))),
+        }
+    }
+}
+
+impl Selector {
+    pub fn find<'a>(&'a self, input: &'a Value) -> Iter<'a> {
+        match self {
+            Selector::Union(indices) => Box::new(indices.iter().flat_map(move |i| i.get(input))),
+            Selector::DotName(name) => Box::new(input.get(name).into_iter()),
+            Selector::DotWildcard => match input {
+                Value::Object(m) => Box::new(m.values()),
+                Value::Array(a) => Box::new(a.iter()),
+                _ => Box::new(std::iter::empty()),
+            },
```

Comment:
I really like the readability of this approach. (It's a pity `Box::new` can't be factored out of the outer match, but that's rust for you.)

Reply:
Hopefully soon rust will have the `box` keyword which at least removes the need for extra parentheses